mosqagent.c

```diff
@@ -105,6 +107,100 @@ void mqtta_dispose_message(struct mqtta_message *msg)
     free(msg);
 }
 
+/**
+ * If there is already a configuration in the agent, make sure to take care
+ * of the memory object before calling this function, otherwise there may be
+ * memory leaks!
+ */
+int mqtta_load_configuration(struct mosqagent *agent,
+                             const char* filepath)
+{
+    int ret = 0;
+
+    struct mosqagent_config *config;
+
+    config_t configuration;
+
+    const char* client_name;
+    const char* broker_host;
+
+    // Init the configuration struct from libconfig
+    config_init(&configuration);
+
+    /*
+     * Check this part first. Not finding the config file is the most likely
+     * failure and we'll fail this before reserving memory.
+     */
+
+    if (!config_read_file(&configuration, filepath)) {
+        fprintf(stderr, "Cannot read config file: %s\n", config_error_text(&configuration));
+        ret = 1;
+        goto cleanup_with_configuration;
+    }
+
+    // Now create the memory object
+    config = malloc(sizeof(*config));
+    if (!config) {
+        ret = -ENOMEM;
+        goto cleanup_with_configuration;
+    }
+
+    // We have to find an agent name!
+    if (config_lookup_string(&configuration, "mosqagent.name", &client_name))
+    {
+        config->client_name = malloc(strlen(client_name));
+        strcpy(config->client_name, client_name);
+    } else {
```

Comment:
The buffer you allocate here must be 1 byte greater, otherwise `strcpy` writes the terminating zero to memory you didn't allocate. You may consider to use `strdup` which is not standard C or better use the safe variants `strncpy` (does not ensure null termination in corner case) or `snprintf`.

Reply:
I fixed this - please have a look at the new code.

Reply:
From the top of my head I would say the additional setting of null termination is redundant, `strncpy` does this, "fails" only if string length equals buffer length, which you prevented by allocating one byte more.